### PR TITLE
refactor ExecutionPath to improve performance and simplify it

### PR DESCRIPTION
**Pull request by andimarek**
Technically breaking change because it is public API.

### CODE REVIEW

File: src/main/java/graphql/execution/ExecutionPath.java [hunk 1]
```diff
@@ -63,27 +68,48 @@ public ExecutionPath getPathWithoutListEnd() {
...
+        return (int) segment;
+    }
+
+    public Object getSegmentValue() {
+        return segment;
+    }
+
+    public ExecutionPath getParent() {
+        return parent;
     }
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ExecutionPath.java:106]: Wow we never had that :)

Comment (andimarek) [reply src/main/java/graphql/execution/ExecutionPath.java:106]: We have drop segment, which is basically the same, right? 
Should we keep drop segment or get parent?

Comment (bbakerman) [reply src/main/java/graphql/execution/ExecutionPath.java:106]: both I think

Comment (bbakerman) [reply src/main/java/graphql/execution/ExecutionPath.java:106]: I think both are ok in that one is a read style thing - tell me the parent - where was dropSegment feels like a write still thing (but its immuatble so its not)

For backwards compatibility AND understanding - I would have both

**Pull request merged**